library/ssl_tls13_server.c

```diff
@@ -97,6 +98,319 @@ static int ssl_tls13_parse_key_exchange_modes_ext( mbedtls_ssl_context *ssl,
+        p += binder_len;
+
+        if( binder_matched == SSL_TLS1_3_OFFERED_PSK_MATCH)
+            continue;
+
+        binder_matched = ssl_tls13_offered_psks_check_binder_match(
+                              ssl, binder, binder_len, psk, psk_len );
+        if( binder_matched < 0 )
+        {
+            MBEDTLS_SSL_DEBUG_RET( 1,
+                "ssl_tls13_offered_psks_check_binder_match" , binder_matched );
+            MBEDTLS_SSL_PEND_FATAL_ALERT(
+                MBEDTLS_SSL_ALERT_MSG_HANDSHAKE_FAILURE,
+                MBEDTLS_ERR_SSL_HANDSHAKE_FAILURE );
+            ret = binder_matched;
+            goto exit_failue;
+        }
+    }
+
+    if( binder_matched != SSL_TLS1_3_OFFERED_PSK_MATCH )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 1,
+            ( "Received psk binder does not match computed psk binder." ) );
+        MBEDTLS_SSL_PEND_FATAL_ALERT(
+            MBEDTLS_SSL_ALERT_MSG_HANDSHAKE_FAILURE,
+            MBEDTLS_ERR_SSL_HANDSHAKE_FAILURE );
+        ret = MBEDTLS_ERR_SSL_HANDSHAKE_FAILURE;
+        goto exit_failue;
+    }
+
+    if( p != binders_end )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 3, ( "pre_shared_key extesion decode error" ) );
+        MBEDTLS_SSL_PEND_FATAL_ALERT( MBEDTLS_SSL_ALERT_MSG_DECODE_ERROR,
+                                      MBEDTLS_ERR_SSL_DECODE_ERROR );
+        ret = MBEDTLS_ERR_SSL_DECODE_ERROR;
+        goto exit_failue;
+    }
+
+    /* Update the handshake transcript with the binder list. */
+    ssl->handshake->update_checksum( ssl,
+                                     identities_end,
+                                     (size_t)( p - identities_end ) );
+    ret = 0;
+
+exit_failue:
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+    mbedtls_free( (void *)psk );
+#endif
+    return( ret );
+}
+
+/*
+ * struct {
+ *   select ( Handshake.msg_type ) {
+ *      ....
+ *      case server_hello:
+ *          uint16 selected_identity;
+ *   }
+ * } PreSharedKeyExtension;
+ */
+static int ssl_tls13_write_selected_identity_ext( mbedtls_ssl_context *ssl,
+                                                  unsigned char *buf,
+                                                  unsigned char *end,
+                                                  size_t *olen )
+{
+    unsigned char *p = (unsigned char*)buf;
+    size_t selected_identity;
+
+    *olen = 0;
+
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+    if( mbedtls_svc_key_id_is_null( ssl->handshake->psk_opaque ) )
+#else
+    if( ssl->handshake->psk == NULL )
+#endif
+    {
+        /* We shouldn't have called this extension writer unless we've
+         * chosen to use a PSK. */
+        return( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
+    }
+
+    MBEDTLS_SSL_DEBUG_MSG( 3, ( "server hello, adding pre_shared_key extension" ) );
+    MBEDTLS_SSL_CHK_BUF_PTR( p, end, 6 );
+
+    MBEDTLS_PUT_UINT16_BE( MBEDTLS_TLS_EXT_PRE_SHARED_KEY, p, 0 );
+    MBEDTLS_PUT_UINT16_BE( 2, p, 2 );
+
+    /* NOTE: This will need to be adjusted once we support multiple PSKs
```

Comment:
I think with the current code we are quite close to supporting multiple PSKs offered by the client thus I'd suggest to finalize it in this PR.

Reply:
Done. Multiple PSK code has been upload. There are too many changes. Should I re-org the commits?

Reply:
no need to re-org the commits for me